File: Array-String/HashTable/CPP/check_array_formation_through_concatenation.cpp
#include <vector>
#include <unordered_map>
using namespace std;
// 1640. Check Array Formation Through Concatenation
// https://leetcode.com/problems/check-array-formation-through-concatenation/description/

class Solution {
public:
    bool canFormArray(vector<int>& arr, vector<vector<int>>& pieces) {
        unordered_map<int, int> idxs;
        int n = arr.size();
        for (int i = 0; i < n; i++) idxs[arr[i]] = i;

        int remain = n;
        for (const vector<int>& piece : pieces) {
            int piece_width = piece.size();
            if (idxs.count(piece[0]) == 0) return false;

            int prev = idxs[piece[0]];
            remain--;

            for (int j = 1; j < piece_width; j++) {
                if (idxs.count(piece[j]) == 0 || idxs[piece[j]] != ++prev)
                    return false;
                remain--;
            }
        }

        return remain == 0;
    }

    // Direct Mapping
    bool canFormArray(vector<int>& arr, vector<vector<int>>& pieces) {
        unordered_map<int, vector<int>> piece_map;

        // Map the first element of each piece to the full piece
        for (const auto& piece : pieces) {
            piece_map[piece[0]] = piece;
        }

        int i = 0;
        while (i < arr.size()) {
            if (piece_map.count(arr[i]) == 0) return false;

            const vector<int>& piece = piece_map[arr[i]];
            for (int num : piece) {
                if (i >= arr.size() || arr[i] != num) return false;
                i++;
            }
        }

        return true;
    }
};

File: Array-String/HashTable/Python/check_array_formation_through_concatenation.py
class Solution:
    def canFormArray(self, arr: list[int], pieces: list[list[int]]) -> bool:
        remain, idxs = len(arr), {v: i for i, v in enumerate(arr)}
        for piece in pieces:
            piece_width = len(piece)
            if piece[0] not in idxs: return False

            prev = idxs[piece[0]]
            remain -= 1

            for j in range(1, piece_width):
                if (piece[j] not in idxs) or (idxs.get(piece[j]) != prev + 1):
                    return False
                prev += 1
                remain -= 1
        return remain == 0
